app/src/main/java/protect/card_locker/ManageGroupCursorAdapter.java

```diff
@@ -0,0 +1,118 @@
+    public void swapCursor(Cursor inputCursor) {
+        super.swapCursor(inputCursor);
+        mIndexCardMap = new HashMap<>();
+        mIsLoyaltyCardInGroupCache = new HashMap<>();
+        mGetGroupCache = new HashMap<>();
+    }
+
+    @Override
+    public void onBindViewHolder(LoyaltyCardListItemViewHolder inputHolder, Cursor inputCursor){
+        LoyaltyCard loyaltyCard = LoyaltyCard.toLoyaltyCard(inputCursor);
+        Boolean overlayValue = mInGroupOverlay.get(loyaltyCard.id);
+        if((overlayValue != null? overlayValue: isLoyaltyCardInGroup(loyaltyCard.id))) {
+            mAnimationItemsIndex.put(inputCursor.getPosition(), true);
+            mSelectedItems.put(inputCursor.getPosition(), true);
+        }
+        mIndexCardMap.put(inputCursor.getPosition(), loyaltyCard.id);
+        super.onBindViewHolder(inputHolder, inputCursor);
+    }
+
+    private List<Group> getGroups(int cardId){
+        List<Group> cache = mGetGroupCache.get(cardId);
+        if(cache != null){
+            return cache;
+        }
+        List<Group> groups = mDb.getLoyaltyCardGroups(cardId);
+        mGetGroupCache.put(cardId, groups);
+        return groups;
+    }
+
+    private boolean isLoyaltyCardInGroup(int cardId){
+        Boolean cache = mIsLoyaltyCardInGroupCache.get(cardId);
+        if(cache != null){
+            return cache;
+        }
+        List<Group> groups = getGroups(cardId);
+        if (groups.contains(mGroup)){
+            mIsLoyaltyCardInGroupCache.put(cardId, true);
+            return true;
+        }
+        mIsLoyaltyCardInGroupCache.put(cardId, false);
+        return false;
+    }
+
+    @Override
+    public void toggleSelection(int inputPosition){
+        super.toggleSelection(inputPosition);
+        Integer cardId = mIndexCardMap.get(inputPosition);
+        if (cardId == null){
+            throw(new RuntimeException("cardId should not be null here"));
+        }
+        Boolean overlayValue = mInGroupOverlay.get(cardId);
+        if (overlayValue == null){
+            mInGroupOverlay.put(cardId, !isLoyaltyCardInGroup(cardId));
+        }else{
+            mInGroupOverlay.remove(cardId);
+        }
+    }
+
+    public boolean hasChanged() {
+        return mInGroupOverlay.size() > 0;
+    }
+
+    public void commitToDatabase(){
+        for(Map.Entry<Integer, Boolean> entry: mInGroupOverlay.entrySet()){
+            int cardId = entry.getKey();
+            List<Group> groups = getGroups(cardId);
+            if(entry.getValue()){
+                groups.add(mGroup);
+            }else{
+                groups.remove(mGroup);
+            }
+            mDb.setLoyaltyCardGroups(cardId, groups);
+        }
+    }
+
+    public void importInGroupState(HashMap<Integer, Boolean> cardIdInGroupMap) {
+        mInGroupOverlay = new HashMap<>();
+        for (Map.Entry<Integer, Boolean> entry: cardIdInGroupMap.entrySet()){
+            mInGroupOverlay.put(entry.getKey(), entry.getValue());
+        }
+    }
```

Comment:
Isn't cardIdInGroupMap valid already? Why not just `mInGroupOverlay = cardIdInGroupMap`?